Add docs page for the new `ReanimatedDrawerLayout` component (#3188)
software-mansion/react-native-gesture-handler: docs/docs/components/reanimated-drawer-layout.mdx

---
id: reanimated-drawer-layout
title: Reanimated Drawer Layout
sidebar_label: Reanimated Drawer Layout
---

import useBaseUrl from '@docusaurus/useBaseUrl';

Cross-platform replacement for the React Native's [DrawerLayoutAndroid](http://reactnative.dev/docs/drawerlayoutandroid.html) component.  
Please refer to [React Native docs](http://reactnative.dev/docs/drawerlayoutandroid.html) for the detailed usage for standard parameters.

### Usage:

To use it, import it in the following way:

```js
import ReanimatedDrawerLayout from 'react-native-gesture-handler/ReanimatedDrawerLayout';
```

## Properties:

### `drawerType`

> accepts values of the `DrawerPosition` enum, defaults to `FRONT`.  

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

> Specifies the way the drawer will be displayed.

Author: I agree, fixed in [3ded966](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/3ded9667d37765dd00263c0140a74d80368aabcd)

- `FRONT` the drawer will be displayed above the content view.
- `BACK` the drawer will be displayed below the content view, revealed by sliding away the content view.
- `SLIDE` the drawer will appear attached to the content view, opening it slides both the drawer and the content view.

| `FRONT`                                               | `BACK`                                               | `SLIDE`                                               |
| ----------------------------------------------------- | ---------------------------------------------------- | ----------------------------------------------------- |
| <img src={useBaseUrl('gifs/new-drawer-front.gif')} /> | <img src={useBaseUrl('gifs/new-drawer-back.gif')} /> | <img src={useBaseUrl('gifs/new-drawer-slide.gif')} /> |

### `edgeWidth`

> width of the invisible draggable area on the edge of the content view, which can be dragged to open the drawer.

Author: Fixed in [3ded966](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/3ded9667d37765dd00263c0140a74d80368aabcd)

### `hideStatusBar`

> a boolean value, when set to `true`, drawer component will use [StatusBar API](http://reactnative.dev/docs/statusbar.html) to hide the OS status bar when the drawer is dragged or idle in the `open` position.

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

### `statusBarAnimation`

> a string, possible values are: `slide`, `none` or `fade`, defaults to `slide`.

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

> a string, with possible values: `slide`, `none` or `fade`. Defaults to `slide`.

Author: I think both are correct. 
Tested both sentences on multiple online punctuation checkers which flagged both as correct, and so did ChatGPT.
I can change it if you prefer the variant with no comma.

May be used in combination with `hideStatusBar` to select the animation used for hiding the status bar.
See [StatusBar API](http://reactnative.dev/docs/statusbar.html#statusbaranimation) docs.

### `overlayColor`

color (default `rgba(0, 0, 0, 0.7)`) of the background overlay on top of the content window when the drawer is `open`.

> This color animates from `0%` to `100%` opacity as drawer animates from `closed` to `open`.

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

### `renderNavigationView`

> a renderer function for the drawer componenet.

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

Provided with a `progress` SharedValue parameter that indicates the progress of drawer opening/closing. Equals `0` when drawer is `closed` and `1` when `open`.

> This can be used by the drawer component to animated its children while the drawer is opening or closing.

Reviewer: First of all, I'd probably merge first two sentences as the second one sounds a bit odd on its own. Beside that 

```
Equals `0` when drawer is `closed` and `1` when `open`.
```

What equals `0`? A function? And let's stay consistent, if we use `closed`, then we should also use `opened`.

Either way, I believe it can be rewritten is a better way 😅

Author: I agree, does something like this work better?

>a renderer function for the drawer component, provided with a `progress` SharedValue parameter that indicates the progress of drawer opening/closing animation.
>- `progress` equals `0` when the drawer is `closed` and `1` when the drawer is `opened`.
>- `progress` can be used by the drawer component to animated its children while the drawer is opening or closing.

Inspired by the `progress` description used in the `reanimated_swipeable` docs.

Reviewer: It is definitely better 😅 What about:

<hr />

a renderer function for the drawer component, provided with a `progress` parameter.

-  `progress` - `SharedValue` that indicates the progress of drawer opening/closing animation.
    - equals `0` when the `drawer` is closed and `1` when the `drawer` is opened
    - can be used by the `drawer` component to animated its children while the `drawer` is opening or closing

Author: I agree, changed in [ca3eda4](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/ca3eda4375625b8578bac3fc09e45fdc36af4977)

<details>
<summary>
Collapsed comparison
</summary>

before:
![image](https://github.com/user-attachments/assets/e31ed3e8-73a2-4fae-ad9e-35dede1cf252)
after:
![image](https://github.com/user-attachments/assets/af6d4f9e-6a5e-4b1b-9f37-87458611e693)
</details>

### `onDrawerClose`

> a function which is called when the drawer is closed.

Reviewer: What about 
```suggestion
a function which is called when the drawer has been closed.
```
?

I think it tells more about the timing of this action.

Author: I agree

### `onDrawerOpen`

> a function which is called when the drawer is opened.

Reviewer: Same as above

Author: Same here

### `onDrawerSlide`

> function. Called when drawer is moving or animating.

Reviewer: I think you've missed changing this part 😅

Author: Yep, fixed in [e108994](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/e1089942751eb7421c067cbdbe8309fed29b3f27)

> Provided with a `progress` SharedValue parameter that indicates the progress of drawer opening/closing. Equals `0` when drawer is `closed` and `1` when `open`.

Author: fixed in [d980b5d](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/d980b5d7226cce802a4925efa65e3247a6e9039e)

### `onDrawerStateChanged`

a function which is called when the status of the drawer changes. It takes two arguments:

- `newState` - interaction state of the drawer. It can be one of the following:
  - `DrawerState.IDLE`
  - `DrawerState.DRAGGING`
  - `DrawerState.SETTLING`
- `drawerWillShow` - `true` when `drawer` started animating to `open` position, `false` otherwise.

### `enableTrackpadTwoFingerGesture` (iOS only)

enables two-finger gestures on supported devices, for example iPads with trackpads.

> If not enabled the gesture will require click + drag, with enableTrackpadTwoFingerGesture swiping with two fingers will also trigger the gesture.

Reviewer: ```suggestion
If not enabled the gesture will require click + drag, with `enableTrackpadTwoFingerGesture` swiping with two fingers will also trigger the gesture.
```

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

> If not enabled the gesture will require click + drag, with `enableTrackpadTwoFingerGesture` swiping with two fingers will also trigger the gesture.

Author: Fixed in [22bd637](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/22bd63725014905dacae66431c874ad1d09aa18a)

### `children`

> component or function. Children is a component which is rendered in the content view.
> May also be a render function which may accept a `progress` SharedValue as a parameter that indicates the animation progress, similarily to the `renderNavigationView` render function.

Reviewer: ```suggestion
either a component that's rendered in the content view or a function. When provided as a function, it accepts a `progress` SharedValue parameter that indicates the animation progress, similar to the [`renderNavigationView`](#rendernavigationview) function.
```

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

Author: fixed in [deff142](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/deff1421ce13ec36e5c3f9546684d0d76947ffa0)

### `mouseButton(value: MouseButton)` (Web & Android only)

allows users to choose which mouse button should handler respond to.
The enum `MouseButton` consists of the following predefined fields:

- `LEFT`
- `RIGHT`
- `MIDDLE`
- `BUTTON_4`
- `BUTTON_5`
- `ALL`

Arguments can be combined using `|` operator, e.g. `mouseButton(MouseButton.LEFT | MouseButton.RIGHT)`. Default value is set to `MouseButton.LEFT`.

### `enableContextMenu(value: boolean)` (Web only)

specifies whether context menu should be enabled after clicking on underlying view with right mouse button. Default value is set to `false`.

## Methods

### `openDrawer(options)`

`openDrawer` accepts an optional `options` parameter, which is an object with the following optional properties:

> `initialVelocity`: number, the initial velocity of the object attached to the spring. Default `0`.
> `animationSpeed`: number, controls speed of the animation. Default `1`.

Reviewer: ```suggestion
- `initialVelocity` - initial velocity of the object attached to the spring. Defaults to `0`.
- `animationSpeed`- controls speed of the animation. Defaults to `1`.
```

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

### `closeDrawer(options)`

`closeDrawer` accepts an optional `options` parameter, which is an object with the following optional properties:

> `initialVelocity`: number, the initial velocity of the object attached to the spring. Default `0`.
> `animationSpeed`: number, controls speed of the animation. Default `1`.

Reviewer: ```suggestion
- `initialVelocity` - initial velocity of the object attached to the spring. Defaults to `0`.
- `animationSpeed`- controls speed of the animation. Defaults to `1`.
```

Author: fixed in [060d2da](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/060d2da676256ed4a08d517427bc93327bb76e2d)

## Example:

> See the [reanimated drawer layout example](https://github.com/software-mansion/react-native-gesture-handler/blob/main/example/src/release_tests/reanimatedDrawerLayout/index.tsx) from GestureHandler Example App.

Reviewer: ```suggestion
See the [reanimated drawer layout example](https://github.com/software-mansion/react-native-gesture-handler/blob/main/example/src/release_tests/reanimatedDrawerLayout/index.tsx) from GestureHandler example app.
```

---
id: reanimated-drawer-layout
title: Reanimated Drawer Layout
sidebar_label: Reanimated Drawer Layout
---

import useBaseUrl from '@docusaurus/useBaseUrl';

Cross-platform replacement for the React Native's [DrawerLayoutAndroid](http://reactnative.dev/docs/drawerlayoutandroid.html) component.  

> Provides a compatible API while allows for the component to be used on both Android and iOS.  

Reviewer: Isn't it also available on `web`?

Please refer to [React Native docs](http://reactnative.dev/docs/drawerlayoutandroid.html) for the detailed usage for standard parameters.

### Usage:

To use it, import it in the following way:

```js
import ReanimatedDrawerLayout from 'react-native-gesture-handler/ReanimatedDrawerLayout';
```

## Properties:

### `drawerType`

accepts values of the `DrawerPosition` enum, defaults to `DrawerPosition.FRONT`.  
Specifies the way the drawer will be displayed.

- `DrawerPosition.FRONT` the drawer will display on top of the content view.
- `DrawerPosition.BACK` the drawer appears to be behind the content view, revealed by sliding away the content view.

> - `DrawerPosition.SLIDE` the drawer displays attached to the content view, opening it slides both the drawer and the content view.

Reviewer: I'd use passive voice, i.e. _**drawer will be displayed**_ etc.

Author: Fixed in [ec0e598](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/ec0e598ee6dba06e004d23521e3e51caf01c28ee)

| `DrawerPosition.FRONT`                            | `DrawerPosition.BACK`                            | `DrawerPosition.SLIDE`                            |
| ------------------------------------------------- | ------------------------------------------------ | ------------------------------------------------- |

> | <img src={useBaseUrl('gifs/drawer-front.gif')} /> | <img src={useBaseUrl('gifs/drawer-back.gif')} /> | <img src={useBaseUrl('gifs/drawer-slide.gif')} /> |

Reviewer 2: One nit here: can we use images from the new implementation instead of the old one 😄?

> | <img src={useBaseUrl('gifs/drawer-front.gif')} /> | <img src={useBaseUrl('gifs/drawer-back.gif')} /> | <img src={useBaseUrl('gifs/drawer-slide.gif')} /> |

Author: https://github.com/user-attachments/assets/a38d85d0-cd2a-41d4-a8c5-2081ef3ccbd8

Done in [a215843](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/a215843c0c37f826d6ba3a440cd35276c95a6652) and [a2a7545](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/a2a754511ccc0eaed399138735fc360ee2da3076)

### `edgeWidth`

> number. Width of the invisible draggable area on the edge of the content view, which can be dragged to open the drawer.

Reviewer: I know that it probably would require more changes in our docs, but I don't like the fact that the line starts with _**number.**_

I think that we could either include type in header (and that requires a bit more work in docs), or do something similar to [this](https://docs.swmansion.com/react-native-gesture-handler/docs/components/reanimated_swipeable#friction) or [that](https://docs.swmansion.com/react-native-gesture-handler/docs/components/reanimated_swipeable#friction).

Author: I understand, fixed this and other instances of types being bluntly stated in [a5e9409](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/a5e9409701e45d9c7961bcec2b0a83e07a79ae43)

### `edgeWidth`

width of the invisible, draggable area on the edge of the content view, which can be dragged to open the drawer.

### `hideStatusBar`

a boolean value. When set to `true`, drawer component will use [StatusBar API](http://reactnative.dev/docs/statusbar.html) to hide the OS status bar when the drawer is dragged or idle in the `open` position.

### `statusBarAnimation`

a string with possible values: `slide`, `none` or `fade`. Defaults to `slide`.
May be used in combination with `hideStatusBar` to select the animation used for hiding the status bar.
See [StatusBar API](http://reactnative.dev/docs/statusbar.html#statusbaranimation) docs.

### `overlayColor`

color of the background overlay on top of the content window when the drawer is `open`.  
This color's opacity animates from 0% to 100% as the drawer transitions from closed to open. Defaults to `rgba(0, 0, 0, 0.7)`.

### `renderNavigationView`

a renderer function for the drawer component, provided with a `progress` parameter.

-  `progress` - `SharedValue` that indicates the progress of drawer opening/closing animation.

>     - equals `0` when the `drawer` is closed and `1` when the `drawer` is opened

Author: Cambridge dictionary says it should be `open` when the word is used as an adjective, and `opened` when it's used as a past/present/future perfect verb.
In this context, it could very well be either a verb, or an adjective, and so both forms are likely valid.

### `children`

either a component that's rendered in the content view or a function.

> When provided as a function, it is provided with a `progress` parameter.

Reviewer: **_Provided_** twice in the same, short sentence doesn't sound good 😅

Author: does `supplied` work for the second `provided`?

Reviewer: What about simply:

```
If `children` is a function, it is provided with a `progress` parameter.
```

Author: Sounds good, fixed in [aa73f47](https://github.com/software-mansion/react-native-gesture-handler/pull/3188/commits/aa73f4741ba03ae51b59e5a30a3cff589acb6212)
